### PR TITLE
8221554: aarch64 cross-modifying code

### PR DESCRIPTION
**Pull request by a74nh**
The AArch64 port uses maybe_isb in places where an ISB might be required
because the code may have safepointed. These maybe_isbs are very conservative
and are used in many places are used when a safepoint has not happened.

cross_modify_fence was added in common code to place a barrier in all the
places after a safepoint has occurred. All the uses of it are in common code,
yet it remains unimplemented on AArch64.

This set of patches implements cross_modify_fence for AArch64 and reconsiders
every uses of maybe_isb, discarding many of them. In addition, it introduces
a new diagnostic option, which when enabled on AArch64 tests the correct
usage of the barriers.

Advantage of this patch is threefold:
* Reducing the number of ISBs - giving a theoretical performance improvement.
* Use of common code instead of backend specific code.
* Additional test diagnostic options

Patch 1: Split cross_modify_fence
=================================
This is simply refactoring work split out to simplify the other two patches.

instruction_fence() is provided by each target and simply places
a fence for the instruction stream.

cross_modify_fence() is now a member of JavaThread and just calls
instruction_fence. This function will be extended in Patch 3.

Patch 2: Use cross_modify_fence instead of maybe_isb
====================================================

The [n] References refer to the comments for cross_modify_fence in
thread.hpp.

This is all the existing uses of maybe_isb in the AArch64 target:

1) Instances of Java code calling a VM function
   * This encapsulates the changes to:
   ** MacroAssembler::call_VM_leaf_base()
   ** generate_fast_get_int_field0()
   ** stubGenerator_aarch64 generate_throw_exception()
   ** sharedRuntime_aarch64 generate_handler_blob()
   ** SharedRuntime::generate_resolve_blob()
   ** C1 LIR_Assembler::rt_call
   ** C1 StubAssembler::call_RT(): used by Used by generate_exception_throw,
                                   generate_handle_exception, generate_code_for.
   ** OptoRuntime::generate_exception_blob()
   * Any changes will be caught due to calls to [2] or [3] by the VM function.
   * Any calls that do not call [2] or [3] do not require an ISB.
   * This patch is more optimal for these cases.

2) Instances of Java code calling a JNI function
   * This encapsulates the changes to:
   ** SharedRuntime::generate_native_wrapper()
   ** TemplateInterpreterGenerator::generate_native_entry()
   * A safepoint still in progress after the call with be caught by [4].
   * An ISB is still required for the case where there was a safepoint
     but it completed during the call. This happens if the code doesn't
     branch on safepoint_in_progress
   * In the SharedRuntime version, the two possible calls to
     reguard_yellow_pages and complete_monitor_unlocking_C are after the thread
     goes back into it's original state, so are covered by [2] and [3], the
     same as a normal VM call.
   * This patch is only more optimal for the two post-JNI calls.

3) Patching functions
   * This encapsulates the changes to:
   ** patch_callers_callsite() (called by gen_c2i_adapter())
   * This results in code being patched, but does not safepoint
   * Therefore an ISB is required.
   * This patch introduces no change here.

4) C1 MacroAssembler::emit_static_call_stub()
   * Calls ISB (not maybe_isb)
   * By design, the patching doesn't require that the up-to-date
     destination is required for proper functioning.
   * However, the ISB makes it most likely that the new destination will
     be picked up.
   * This patch introduces no change here.

Patch 3: Add cross modify fence verification
============================================

The VerifyCrossModifyFence diagnostic flag enables confirmation to the correct
usage of instruction barriers. It can safely be enabled on any Java run.

Enabling it will cause the following:

* Once all threads have been brought to a safepoint, each thread will be
  marked.

* On a cross_modify_fence and safepoint_fence the mark for that thread
  will be cleared.

* On entry to a method and in a safepoint poll, then the thread is checked.
  If it is marked, then the code will error.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8221554](https://bugs.openjdk.java.net/browse/JDK-8221554): aarch64 cross-modifying code

### Reviewers
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to 338eca42fde713cbab70cf7fd15d9fad5aab7724
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/428/head:pull/428`
`$ git checkout pull/428`

### CODE REVIEW

**Comment by fisk** [comment]
I am about to integrate concurrent stack scanning for ZGC. This patch changes things a bit so that disarming of the poll word is always done only by the thread itself. The consequence is that if a thread is in native or blocked, and the safepoint finishes, then the thread will when waking up, still take one slow path to figure out all is done, disarm the poll and run the cross modifying fence. The consequence is that we trivially need the cross modifying fence only in the poll slow path and nowhere else. So maybe if you wait for that one, we can delete a few more ISBs, and IMO also delete the verification code, as you can't really mess it up any more.

**Comment by a74nh** [comment]
> I am about to integrate concurrent stack scanning for ZGC. This patch changes things a bit so that disarming of the poll word is always done only by the thread itself. The consequence is that if a thread is in native or blocked, and the safepoint finishes, then the thread will when waking up, still take one slow path to figure out all is done, disarm the poll and run the cross modifying fence. The consequence is that we trivially need the cross modifying fence only in the poll slow path and nowhere else. So maybe if you wait for that one, we can delete a few more ISBs, and IMO also delete the verification code, as you can't really mess it up any more.

I'd need to see what the code looks like - but I'm thinking it just reduces the number of cross_modify_fence calls, instead of reducing explicit isb calls in the backend (of which only 3 remain with this patch). Either way, it's good news. I'm going to try rebasing locally on the top of your patch (I'm assuming it's this one https://github.com/openjdk/jdk/pull/296 ) and see where that gets me.

I'd be cautious about removing the verification code - it should be useful for helping to catch issues if we do start getting  1 run in a million crash dumps. But agreed, doesn't need to be there if the whole isb process has become trivial.

**Comment by fisk** [comment]
You no longer need any isb when returning from the native wrappers (interpreted or compiled variant) after my patch. That *should* be 2 if the mentioned 3 hooks (without looking closely at the code). Because that will be done in the runtime instead when waking up from native. Which hook do we have left?

**Comment by a74nh** [comment]
Agreed. With both patches, and with the JNI isbs removed, that leaves just one safepoint isb in the AArch64 code (plus the other isb in emit_static_call_stub).

The test patch exists not just to test the AArch64 but the common code too (ideally it would be extended to other targets too). Are we happy that the cross_modify_fence is called at all the required points? A hole in the common code would fail only very rarely. It does require quite a bit of code to add this test though.

**Comment by robehn** [comment]
The only reason for having cmf() on thread is the validation in debug builds, right?
If you do it the opposite way:

inline void OrderAccess::cross_modify_fence_non_arch_specific() {
    OrderAccess::cross_modify_fence_arch_impl();
#ifdef ASSERT
    if (VerifyCrossModifyFence) {
      Thread::current()->set_requires_cross_modify_fence(false);
    }
#endif
}

You only need the boolean in debug builds on JavaThreads and you don't need to move the cmf() from OA and create the new one?

**Comment by a74nh** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> Hi Alan,
> 
> On 1/10/2020 2:30 am, Alan Hayward wrote:
> 
> > The AArch64 port uses maybe_isb in places where an ISB might be required
> > because the code may have safepointed. These maybe_isbs are very conservative
> > and are used in many places are used when a safepoint has not happened.
> > cross_modify_fence was added in common code to place a barrier in all the
> > places after a safepoint has occurred. All the uses of it are in common code,
> > yet it remains unimplemented on AArch64.
> > This set of patches implements cross_modify_fence for AArch64 and reconsiders
> > every uses of maybe_isb, discarding many of them. In addition, it introduces
> > a new diagnostic option, which when enabled on AArch64 tests the correct
> > usage of the barriers.
> > Advantage of this patch is threefold:
> > * Reducing the number of ISBs - giving a theoretical performance improvement.
> > * Use of common code instead of backend specific code.
> > * Additional test diagnostic options
> > Patch 1: Split cross_modify_fence
> > =================================
> > This is simply refactoring work split out to simplify the other two patches.
> > instruction_fence() is provided by each target and simply places
> > a fence for the instruction stream.
> > cross_modify_fence() is now a member of JavaThread and just calls
> > instruction_fence. This function will be extended in Patch 3.
> 
> I don't agree with the change here. The cross_modify_fence() is not
> related to thread API imo, it belongs in OrderAccess. The name was
> deliberately selected to abstract away from the specific details of why
> a given platform may need this fence:
> 
> http://mail.openjdk.java.net/pipermail/hotspot-dev/2019-March/037153.html
> 
> "The name "instruction_pipeline" seems a bit implementation specific
> about what HW architectural features need to be taken care of due to
> cross-modifying code, which may or may not apply to a given platform.
> Perhaps cross_modify_fence(), or something along those lines, would be
> better. That makes it more clear what we are protecting against, as
> opposed to what HW architectural features that might concern on a given
> platform."
> 
> @robehn , @fisk please chime in here. :)
> 
> Thanks,
> David

I have no strong feeling on the names of these functions.
The reason for moving it was that in the third part (the verification testing) it needs JavaThread.
In an earlier version I simply had OrderAccess::cross_modify_fence(JavaThread *thread). But then OrderAccess is dependant on JavaThread, which felt wrong. Obvious solution was to add a wrapper in JavaThread that calls down to OrderAccess.
Alternatively, I could switch the name of instruction_fence back to cross_modify_fence, and then think of a name for the added function in JavaThread. Alternatively alternatively, they could both be call cross_modify_fence.
If the verification test patch was removed from the set, then most of the first patch wouldn't be needed either.

**Comment by a74nh** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 02/10/2020 09:08, Alan Hayward wrote:
> 
> > I have no strong feeling on the names of these functions.
> > The reason for moving it was that in the third part (the verification testing) it needs JavaThread.
> 
> Right, but you can get the JavaThread efficiently any time you want,
> so you don't need to pass it to cross_modify_fence().
> 

Oh, ok, didn't spot that.
This would result in code in OrderAccess.cpp calling a function in JavaThread.
It feels that OrderAccess should be much lower level than JavaThread. But, that might be ok.

**Comment by a74nh** [comment]
Patch updated.

* cross_modify_fence now calls cross_modify_fence_impl as suggested.

* ISBs in the JNI calls have been removed. This means that it is currently unsafe to merge until https://github.com/openjdk/jdk/pull/296 has been merged.

**Comment by a74nh** [comment]
https://github.com/openjdk/jdk/pull/296 has been merged to head.

Rebased this request and re-tested.
All review comments have been addressed.
Note that with VerifyCrossModifyFence set, testing took longer by a factor of *2.5.

**Comment by dholmes-ora** [comment]
@a74nh Please do not force-push commits on an open PR as it breaks the commit history and prevents reviewers from seeing what has changed since they last reviewed things. If you need to "rebase" you can just merge your branch with an updated master branch and push the merge commit to your personal fork. The skara tooling will flatten the commits into a single clean commit when integration happens. Thanks.

**Comment by a74nh** [comment]
> @a74nh Please do not force-push commits on an open PR as it breaks the commit history and prevents reviewers from seeing what has changed since they last reviewed things. If you need to "rebase" you can just merge your branch with an updated master branch and push the merge commit to your personal fork. The skara tooling will flatten the commits into a single clean commit when integration happens. Thanks.

Not a fan of working with merge commits and I feel it gets muddled when you have history on top of a patch series (as opposed to a single patch). However, understood - I'll make sure to merge instead of force pushing next time.

File: src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp:35]: This #define of inlasm_isb() looks wrong. Surely it should be in the body of OrderAccess::cross_modify_fence_impl() given that it's not used anywhere else. All the extra indirection does is confuse the reader.

Comment (a74nh) [reply src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp:35]: Agreed. It was designed to fit with my patch which did the same for the dmb's - but I've closed that patch. Will fix.

Comment (a74nh) [reply src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp]: Updated

**Comment by a74nh** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> So, the good news and the bad news:
> 
> Moving to cross_modify_fence reduces the number of ISBs from
> 3,840,210 maybe_isb()s to 74,538 cross_modify_fence()s on my
> poster child application, which is recompiling all of java.base.
> 
> However, this is a program that runs for 187,501,798,979 insns,
> so we've reduced the proportion of ISBs from 0.002% to 0.00004%.
> I guess that's worth having, but I doubt that the improvement
> would ever have been above the noise level.

Thanks for testing this out.
"Kills 98% of ISBs" is the marketing headline then.
Sadly less effective than I was hoping - which explains my testing results.

> 
> On the good side, this at least makes AArch64 more like other
> targets.
> 

Yes. This should give all the goodness from using common code
(simpler, stronger code, etc etc)

To be clear, the four original reasons for the patch were:
*Use common code/interfaces where possible
*Reduce ISBs where AArch64 was being too cautious
*Add ISBs if theres any paths without them   (there weren't)
*Confirm the above changes are safe.

**Comment by pchilano** [comment]
Regarding the use of cross_modify_fence(), I filed a bug last week to remove some uneeded uses of them in common code (https://bugs.openjdk.java.net/browse/JDK-8254264). Just a heads up before I send the RFR since I see some reference to them in the added comments.

Thanks,
Patricio

**Comment by a74nh** [comment]
> Regarding the use of cross_modify_fence(), I filed a bug last week to remove some uneeded uses of them in common code (https://bugs.openjdk.java.net/browse/JDK-8254264). Just a heads up before I send the RFR since I see some reference to them in the added comments.
> 

I'm going to assume your change is just a two line change (removing the cross_modify_fence's), and I'll test that on top of my patches using the VerifyCrossModifyFence flag - I'll give it a run of everything, which can take a while. Plus I'll manually look at the code to to make sure I'm happy.

I think it makes sense that your patch goes in first, then I can rebase and update code comments too. Let me know your pull request once you've raised it.

**Comment by pchilano** [comment]
Yes, the change just removes those extra cross_modify_fence's. Please check https://github.com/openjdk/jdk/pull/655

**Comment by robehn** [comment]
A question, ISB don't flush the I-cache which I thought was needed?
I would have expected something more similar to gcc clear_cache.

**Comment by a74nh** [comment]
> A question, ISB don't flush the I-cache which I thought was needed?
> I would have expected something more similar to gcc clear_cache.

(Possible I've missed something in your question...)
Any cache flushing would be performed by the thread that modifies the code.
The cross_modify_fence is for the other threads after the code has been modified. The only thing they need to do is an isb.

**Comment by robehn** [comment]
> > A question, ISB don't flush the I-cache which I thought was needed?
> > I would have expected something more similar to gcc clear_cache.
> 
> (Possible I've missed something in your question...)
> Any cache flushing would be performed by the thread that modifies the code.
> The cross_modify_fence is for the other threads after the code has been modified. The only thing they need to do is an isb.

Correct me if I'm wrong:
We write the updated instructions to D-cache which is not coherent with the I-cache.
To synchronize the I-cache to what we wrote in the D-cache, we first need make sure the D-cache is committed to memory.
Then each processor must invalidate it's I-cache, thus fetching the new instructions from memory.
So yes the thread that modifies the code should make sure the D-cache is consistent with main memory. 
Can the I-cache invalidation be done by the modifying thread in a distributed way ?

When we do "fix_oop_relocations" we end up calling: MacroAssembler::patch_oop(..) in aarch64 case.
To me this indicate that we just changed an oop in a code segment, but I see no signs of above.

Maybe I got something wrong about how this I-cache/D-cache works or missed something in our code?

I think @theRealAph suggested having a indirect to the oops and thus move them of out code and avoiding I-cache.

**Comment by theRealAph** [comment]
On 15/10/2020 17:10, Robbin Ehn wrote:
>
> Correct me if I'm wrong:
> We write the updated instructions to D-cache which is not coherent with the I-cache.
> To synchronize the I-cache to what we wrote in the D-cache, we first need make sure the D-cache is committed to memory.
Yes.

> Then each processor must invalidate it's I-cache, thus fetching the new instructions from memory.
> So yes the thread that modifies the code should make sure the D-cache is consistent with main memory.
> Can the I-cache invalidation be done by the modifying thread in a distributed way ?

Yes, it is. In a multiprocessor system, IC IVAU instruction is broadcast to all cores in the same memory domain.

> When we do "fix_oop_relocations" we end up calling: MacroAssembler::patch_oop(..) in aarch64 case.
> To me this indicate that we just changed an oop in a code segment, but I see no signs of above.

The call stack looks like this:

#0 ICache::invalidate_range  /home/aph/jdk-jdk/src/hotspot/os_cpu/linux_aarch64/icache_linux_aarch64.hpp:40
#1 Relocation::pd_set_data_value  /home/aph/jdk-jdk/src/hotspot/cpu/aarch64/relocInfo_aarch64.cpp:58
#2 DataRelocation::set_value  /home/aph/jdk-jdk/src/hotspot/share/code/relocInfo.hpp:849
#3 DataRelocation::set_value  /home/aph/jdk-jdk/src/hotspot/share/code/relocInfo.hpp:844
#4 oop_Relocation::fix_oop_relocation  /home/aph/jdk-jdk/src/hotspot/share/code/relocInfo.cpp:554
#5 nmethod::fix_oop_relocations  /home/aph/jdk-jdk/src/hotspot/share/code/nmethod.cpp:1051

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

**Comment by robehn** [comment]
> #0 ICache::invalidate_range

Thanks for the pointer!

**Review by robehn — APPROVED** [review]
Seems fine to me, mostly look at shared code part.

**Comment by a74nh** [comment]
Patch merged to master - so that it's on top of pchilano's patch.
Already tested that both the patches work fine together (did a complete run of all our tests with VerifyCrossModifyFence set to true).
robehn has reviewed this patch, but I think I need a second review too (?)

**Comment by robehn** [comment]
> Patch merged to master - so that it's on top of pchilano's patch.
> Already tested that both the patches work fine together (did a complete run of all our tests with VerifyCrossModifyFence set to true).
> robehn has reviewed this patch, but I think I need a second review too (?)

I would recommend someone that knows aarch64 better than me, as I said mostly looked at shared parts.

File: src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp
Comment (nick-arm) [inline src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp:60]: Don't need this `#undef` any more.

Comment (a74nh) [reply src/hotspot/os_cpu/linux_aarch64/orderAccess_linux_aarch64.hpp]: Missed that, thanks. Removed now

**Review by theRealAph — CHANGES_REQUESTED** [review]
Still pending.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (a74nh) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5335]: Aha - Looks like your comments hadn't been made public until now.

The problem is it massively slows down a run. A tier1 test run for fastdebug went from 1h 32m 58s to 
3h 43m 47s. I didn't think that would be acceptable.

File: src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp
Comment (a74nh) [reply src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp:1413]: I wanted to avoid mentioning code that no longer exists. (Maybe it's best to just drop the comment?)

How about:

// When we return from the VM, the instruction stream may have
// been modified. Therefore needs an isb is required. The VM will
// have already done this by calling cross_modify_fence().

File: src/hotspot/share/runtime/orderAccess.hpp
Comment (a74nh) [reply src/hotspot/share/runtime/orderAccess.hpp:237]: This better?

// Finally, we define an "instruction_fence" operation, which ensures that all
// instructions that come after the ISB instruction in program order are fetched
// from the cache or memory after the ISB instruction has completed
// It should be used in conjunction with safepointing to ensure that changes
// to the instruction stream are seen on exit from a safepoint. Namely:
.....etc

Comment (theRealAph) [reply src/hotspot/share/runtime/orderAccess.hpp:237]: // Finally, we define an "instruction_fence" operation, which ensures that all
// instructions that come after the fence in program order are fetched
// from the cache or memory after the fence has completed

**Comment by a74nh** [comment]
> 
> > The problem is it massively slows down a run. A tier1 test run for fastdebug went from 1h 32m 58s to
> > 3h 43m 47s. I didn't think that would be acceptable.
> 
> But why is it so expensive? All it does is mark the threads at a safepoint
> and later check the mark at safepoints. It's not as if it's doing anything
> much, but you're telling me it is more expensive than everything else put
> together.
> 

Ok, please ignore my comment about slowdowns.
Turns out that our testing infrastructure had been scheduling onto different boxes. Once forced onto a single machine, I see no real difference in a complete tier 1 run with and without VerifyCrossModifyFence. (With the slower boxes taking ~4 hours and a quicker one taking ~1 hours)

Patch updated to enable VerifyCrossModifyFence always for aarch64 debug.
 
> > I wanted to avoid mentioning code that no longer exists. (Maybe it's best to just drop the comment?)
> 
> The comment only makes sense in the context of the code that was there
> before.
> 
> > How about:
> > // When we return from the VM, the instruction stream may have
> > // been modified. Therefore needs an isb is required. The VM will
> > // have already done this by calling cross_modify_fence().
> 
> This is self contradicting: firstly you say and ISB is required, then
> you say why it isn't.

Agreed. The comment only makes sense when it refers to code that used to exist. And I really dislike comments that do that. So, I've dropped them completely.

> // Finally, we define an "instruction_fence" operation, which ensures that all
> // instructions that come after the fence in program order are fetched
> // from the cache or memory after the fence has completed

Done.

**Review by mo-beck — CHANGES_REQUESTED** [review]
Could you please incorporate the following orderAccess changes for windows_aarch64.hpp in your PR?: https://gist.github.com/mo-beck/2dc66d068741030b4a422b57607bea8e#file-orderaccess_windows_aarch64_hpp-diff

**Comment by a74nh** [comment]
> Could you please incorporate the following orderAccess changes for windows_aarch64.hpp in your PR?: https://gist.github.com/mo-beck/2dc66d068741030b4a422b57607bea8e#file-orderaccess_windows_aarch64_hpp-diff

Happy to do this. The file has been added since I started the patch, so that's fine. But as a disclaimer for myself, I've no way of testing it, and I'm going to have to assume it just works for Windows AArch64.

**Review by theRealAph — APPROVED** [review]

**Pull request closed without merging**
